silkworm/db/snapshots/bittorrent/web_session.cpp

```diff
@@ -0,0 +1,128 @@
+/*
+   Copyright 2024 The Silkworm Authors
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+#include "web_session.hpp"
+
+#include <map>
+
+#include <boost/asio/ip/tcp.hpp>
+#include <boost/asio/ssl/error.hpp>
+#include <boost/asio/ssl/stream.hpp>
+#include <boost/beast/core/flat_buffer.hpp>
+#include <boost/beast/core/tcp_stream.hpp>
+#include <boost/beast/ssl.hpp>
+#include <boost/beast/version.hpp>
+#include <boost/url/parse.hpp>
+
+#include <silkworm/infra/common/log.hpp>
+
+#include "root_certificates.hpp"
+
+namespace silkworm::snapshots::bittorrent {
+
+namespace net = boost::asio;
+namespace ssl = net::ssl;
+namespace urls = boost::urls;
+
+//! Custom HTTP headers to include in any request to web servers hosted by Cloudflare
+const std::map<std::string_view, std::string_view> kCloudflareHeaders{
+    {"lsjdjwcush6jbnjj3jnjscoscisoc5s", "I%OSJDNFKE783DDHHJD873EFSIVNI7384R78SSJBJBCCJBC32JABBJCBJK45"},
+};
+
+//! The timeout for HTTP asynchronous operations
+constexpr std::chrono::seconds kHttpTimeoutSecs{30};
+
+WebSession::WebSession(std::optional<std::string> server_certificate)
+    : server_certificate_(std::move(server_certificate)) {}
+
+Task<WebSession::StringResponse> WebSession::https_get(const urls::url& web_url, std::string_view target_file) const {
+    // The SSL context which holds root certificate used for verification ()
+    ssl::context ssl_ctx{ssl::context::tlsv13_client};
+    load_root_certificates(ssl_ctx, server_certificate_);
+    ssl_ctx.set_verify_mode(ssl::verify_peer);  // Ask to verify the remote server certificate
+
+    // These objects perform our I/O
+    net::ip::tcp::resolver resolver{co_await net::this_coro::executor};
+    beast::ssl_stream<beast::tcp_stream> ssl_stream{co_await net::this_coro::executor, ssl_ctx};
+    beast::tcp_stream& tcp_stream = beast::get_lowest_layer(ssl_stream);
+
+    // Set SNI Hostname (many hosts need this to handshake successfully)
+    const std::string host{web_url.host()};
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wold-style-cast"
```

Comment:
why is this needed?

Reply:
It's probably a leftover, I will remove it

Reply:
It's needed because of the `SSL_set_tlsext_host_name` signature, see failure [here](https://app.circleci.com/pipelines/github/erigontech/silkworm/12776/workflows/cbfb0a35-3e31-4af4-b77a-8d535e848b66/jobs/56511)